provider/datagen/src/transform/cldr/datetime/symbols.rs

```diff
@@ -86,64 +86,109 @@ fn get_month_code_map(calendar: &str) -> &'static [TinyStr4] {
     }
 }
 
-fn get_era_code_map(calendar: &str) -> BTreeMap<String, TinyStr16> {
-    match calendar {
-        "gregory" => [
-            ("0".to_string(), tinystr!(16, "bce")),
-            ("1".to_string(), tinystr!(16, "ce")),
-        ]
-        .into_iter()
-        .collect(),
-        "buddhist" => [("0".to_string(), tinystr!(16, "be"))]
-            .into_iter()
-            .collect(),
-        "chinese" => [].into_iter().collect(),
-        "japanese" | "japanext" => crate::transform::cldr::calendar::japanese::get_era_code_map(),
-        "coptic" => [
-            // Before Diocletian
-            ("0".to_string(), tinystr!(16, "bd")),
-            // Anno Diocletian/After Diocletian
-            ("1".to_string(), tinystr!(16, "ad")),
-        ]
-        .into_iter()
-        .collect(),
-        "dangi" => [].into_iter().collect(),
-        "indian" => [("0".to_string(), tinystr!(16, "saka"))]
-            .into_iter()
-            .collect(),
-        "islamic" => [("0".to_string(), tinystr!(16, "islamic"))]
-            .into_iter()
-            .collect(),
-        "islamicc" => [("0".to_string(), tinystr!(16, "islamic"))]
-            .into_iter()
-            .collect(),
-        "umalqura" => [("0".to_string(), tinystr!(16, "islamic"))]
-            .into_iter()
-            .collect(),
-        "tbla" => [("0".to_string(), tinystr!(16, "islamic"))]
-            .into_iter()
-            .collect(),
-        "persian" => [("0".to_string(), tinystr!(16, "ah"))]
-            .into_iter()
-            .collect(),
-        "hebrew" => [("0".to_string(), tinystr!(16, "hebrew"))]
-            .into_iter()
-            .collect(),
-        "ethiopic" => [
-            ("0".to_string(), tinystr!(16, "incar")),
-            ("1".to_string(), tinystr!(16, "pre-incar")),
-            ("2".to_string(), tinystr!(16, "mundi")),
-        ]
-        .into_iter()
-        .collect(),
-        "roc" => [
-            ("0".to_string(), tinystr!(16, "roc-inverse")),
-            ("1".to_string(), tinystr!(16, "roc")),
+fn get_era_code_map(calendar: &str) -> &'static BTreeMap<String, TinyStr16> {
+    static MAP: once_cell::sync::OnceCell<BTreeMap<String, BTreeMap<String, TinyStr16>>> =
```

Comment:
We can't really do static caching, `DatagenProvider`s can be mutated and are not singletons, this might bite us later. We have some caches on `CldrCache`, which itself is immutable.

Reply:
Yeah but this data _is_ static. There's no way to affect this via changing CLDRCache

(The japanese era data used for symbols is static; it's _verified_ by datagen but it's statically loaded)

Reply:
Ah right. Wouldn't it suffice to cache the result of `crate::transform::cldr::calendar::japanese::compute_era_code_map`? This is parsing the japanese file even if only gregorian is requested. Also, this doesn't need to return maps, the only call site just iterates over the pairs, so building `BTreeMap`s is overkill. For everything except japanese this can even be completely allocation-free if you pass a `Fn((&str, TinyAsciiStr<16>))` closure, and then japanese can be cached only if it's requested.

Reply:
This is being reused in https://github.com/unicode-org/icu4x/pull/4192 . We might be able to clean this up, I don't want to do that now; ideally we can do a second cleanup pass once the old symbols datagen code is removed.

Reply:
There's a _lot_ of scope of improving this, I'm trying to incrementally improve its performance given that I'm writing code that will end up calling this map 6 times as often as we do today.

Reply:
#4192 also only iterates over the pairs. You're making this less efficient in the case where japanese is not requested. I don't see why we can't keep the match and only cache the japanese branch?

Reply:
Ah, fair.

Reply:
Oh, I remember now, I didn't want to allocate the small BTreeMaps a ton of times.

We can switch to a different model without the BTreeMaps but I'd like to do that later

Reply:
I'll try a little bit

Reply:
Got it to return an iterator

Reply:
nice solution!